File: codeforces/1900~1999/1923/e.cc
#include <bits/stdc++.h>

using namespace std;

using i64 = long long;

void solve() {
  int n = 0;
  cin >> n;
  vector<int> c(n+1);
  for (int i = 1; i <= n; ++i) cin >> c[i];

  vector<vector<int>> adj(n+1);
  for (int i = 1; i < n; ++i) {
    int u = 0, v = 0;
    cin >> u >> v;
    adj[u].push_back(v);
    adj[v].push_back(u);
  }

  vector<map<int,i64>> dp(n+1);
  i64 ans = 0;
  function<void(int,int)> dfs = [&](int node, int p) {
    int current_color = c[node];

    int size = 0;
    int largest = -1;
    for (int next_node : adj[node]) if (next_node != p) {
      dfs(next_node, node);
      if ((int)dp[next_node].size() > size) {
        largest = next_node;
        size = dp[next_node].size();
      }
    }

    for (int next_node : adj[node]) if (next_node != p) {
      if (next_node == largest) {
        continue;
      }

      for (auto& [c, count] : dp[next_node]) {
        if (c != current_color) {
          ans += count * dp[largest][c];
        }
        dp[largest][c] += count;
      } 
    }

    if (largest != -1) {
      swap(dp[node], dp[largest]);
    }
    ans += dp[node][current_color];
    dp[node][current_color] = 1;
  };

  dfs(1, -1);

  cout << ans << endl;
}

int main(void) {
  ios::sync_with_stdio(false);
  cin.tie(0);

  int T = 0;
  cin >> T;
  for (int t = 0; t < T; ++t) {
    solve();
  }
  return 0;
}

File: codeforces/1900~1999/1923/e.rs
use std::io::{stdin, stdout, BufWriter, Write};
use std::str::FromStr;
use std::fmt::Debug;
use std::collections::BTreeMap;

fn read_int<T: FromStr>() -> T where <T as FromStr>::Err: Debug {
    let mut s = String::new();
    stdin().read_line(&mut s).unwrap();
    s.trim().parse::<T>().expect("invalid input")
}

fn read_vec<T: FromStr>() -> Vec<T> where <T as FromStr>::Err: Debug {
    let mut s = String::new();
    stdin().read_line(&mut s).unwrap();
    s.split(" ").map(|x| x.trim().parse().expect("invalid input")).collect()
}

fn dfs(node: usize, p: usize, c: &Vec<i32>, adj: &Vec<Vec<usize>>, ans: &mut i64, dp: &mut Vec<BTreeMap<i32, i64>>) {
    let mut largest = 0;
    let mut largest_len = 0;
    for next_node in adj[node].iter() {
        if *next_node == p {
            continue;
        }

        dfs(*next_node, node, c, adj, ans, dp);
        if dp[*next_node].len() > largest_len {
            largest = *next_node;
            largest_len = dp[*next_node].len();
        }
    }

    if largest > 0 {
        dp.swap(node, largest);
    }

    let current_color = c[node];
    for next_node in adj[node].iter() {
        if *next_node == p {
            continue;
        }
        if *next_node == largest {
            continue;
        }

        let c1 = &dp[*next_node];
        let c2 = &mut dp[largest];
        for (color, count) in c1.iter() {
            if *color != current_color {
                if let Some(count2) = c2.get(color) {
                    *ans += count2 * *count;
                }
            }
            c2.entry(*color);
            // dp[largest].entry(color).and_modify(|c| *c += count).or_insert(count);
        }
    }
    if let Some(count) = dp[node].get(&current_color) {
        *ans += count;
    }
    dp[node].entry(current_color).and_modify(|c| *c = 1).or_insert(1);
}

fn solve() {
    let stdout = stdout();
    let mut writer = BufWriter::new(stdout.lock());
    let n = read_int::<usize>();
    let mut c = read_vec::<i32>();
    c.insert(0, 0);

    let mut adj: Vec<Vec<usize>> = vec![Vec::new(); n+1];
    for _i in 1..n {
        let edge = read_vec::<usize>();
        let u = edge[0];
        let v = edge[1];
        adj[u].push(v);
        adj[v].push(u);
    }

    let mut ans: i64 = 0;
    let mut dp: Vec<BTreeMap<i32,i64>> = vec![BTreeMap::new(); n+1];
    dfs(1, 0, &c, &adj, &mut ans, &mut dp);

    writeln!(writer, "{}", ans).ok();
}

fn main() {
    let n = read_int::<i32>();
    for _t in 0..n {
        solve();
    }
}
